### PR TITLE
[Sprint 1] Feature Logic Login

### PR DESCRIPTION
**Pull request by tantuan127**

### CODE REVIEW

File: .idea/modules.xml [hunk 1]
```diff
@@ -4,6 +4,7 @@
     <modules>
       <module fileurl="file://$PROJECT_DIR$/FOrder.iml" filepath="$PROJECT_DIR$/FOrder.iml" />
       <module fileurl="file://$PROJECT_DIR$/android_forder_01.iml" filepath="$PROJECT_DIR$/android_forder_01.iml" />
+      <module fileurl="file://$PROJECT_DIR$/android_forder_01.iml" filepath="$PROJECT_DIR$/android_forder_01.iml" />
```
Comment (daolq3012) [inline .idea/modules.xml]: remove

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 2]
```diff
@@ -25,4 +57,66 @@ public void onStop() {
...
+        //TODO Login Error !
+        Toast.makeText(mContext, R.string.login_error, Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+    public void onLoginSuccess() {
+        //TODO Login Success !
+        Toast.makeText(mContext, R.string.login_success, Toast.LENGTH_SHORT).show();
+        mContext.startActivity(new Intent(mContext, ChooseDomainActivity.class));
+        ((Activity) (mContext)).finish();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: using Navigator please ask ``Tri'``

File: app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java [hunk 3]
```diff
@@ -1,16 +1,28 @@
...
+import rx.subscriptions.CompositeSubscription;
+
 /**
  * Listens to user actions from the UI ({@link LoginActivity}), retrieves the data and updates
  * the UI as required.
  */
 final class LoginPresenter implements LoginContract.Presenter {
-    private static final String TAG = LoginPresenter.class.getName();
-
+    private static final String PASSWORD = "123456";
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java]: remove. It unnecessary

File: app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java [hunk 4]
```diff
@@ -19,5 +31,41 @@ public void onStart() {
...
+        if (!validateDataInput(userName, passWord)) {
+            return;
+        }
+        Subscription subscription = mUserRepository.login(userName, passWord)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(new Action1<User>() {
+                    @Override
+                    public void call(User user) {
+                        if (!userName.equals(user.getEmail()) && !passWord.equals(PASSWORD)) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java]: remove this line, if login error, Server will response error.

File: app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java [hunk 5]
```diff
@@ -19,5 +31,41 @@ public void onStart() {
...
+                            mViewModel.onLoginError();
+                            return;
+                        }
+                        mViewModel.onLoginSuccess();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java]: need save Object User into local in here. It user screens later

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 6]
```diff
@@ -37,4 +67,68 @@ public void onLoginButtonClicked(View view) {
...
+    }
+
+    @Override
+    public void onLoginSuccess() {
+        //TODO Login Success !
+        Toast.makeText(mContext, R.string.login_success, Toast.LENGTH_SHORT).show();
+        Intent intent = new Intent(mContext, ChooseDomainActivity.class);
+        new Navigator(mContext).startActivity(intent);
+
+        ((Activity) (mContext)).finish();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: using ``Navigator``

File: app/src/main/java/com/framgia/forder/utils/binding/BindingUtils.java [hunk 7]
```diff
@@ -36,6 +37,11 @@ public static void setViewPagerTabs(final TabLayout tabLayout, final ViewPager v
         tabLayout.setupWithViewPager(viewPager, true);
     }
 
+    @BindingAdapter("app:errorText")
+    public static void setErrorText(TextInputLayout layout, String text) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/utils/binding/BindingUtils.java]: remove this method. it not used

File: app/src/main/java/com/framgia/forder/utils/navigator/Navigator.java [hunk 8]
```diff
@@ -29,8 +30,8 @@
     private Activity mActivity;
     private Fragment mFragment;
 
-    public Navigator(@NonNull Activity activity) {
-        mActivity = activity;
+    public Navigator(@NonNull Context activity) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/utils/navigator/Navigator.java]: Please keep origin, not change

File: app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java [hunk 9]
```diff
@@ -19,5 +30,47 @@ public void onStart() {
...
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(new Action1<User>() {
+                    @Override
+                    public void call(User user) {
+                        mViewModel.onLoginSuccess();
+                    }
+                }, new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable throwable) {
+                        mViewModel.onLoginError();
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java]: should show errorr messsage for user

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 10]
```diff
@@ -1,20 +1,55 @@
...
+    }
+
+    @Override
+    public void onLoginClick() {
+
+        if (!mPresenter.validateDataInput(mUsername, mPassword)) {
+            return;
+        }
+        mPresenter.login(mUsername, mPassword);
+        mPresenter.saveUser(mUser);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: Only  save user when login successfully

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 11]
```diff
@@ -32,9 +68,76 @@ public void setPresenter(LoginContract.Presenter presenter) {
...
 
-    public void onLoginButtonClicked(View view) {
-        //TODO change later
-        mNavigator.startActivity(MainActivity.class);
-        mNavigator.finishActivity();
+    @Override
+    public void onLoginError() {
+        //TODO Login Error !
+        AlertDialog.Builder dlgAlert = new AlertDialog.Builder(mContext);
+        dlgAlert.setMessage(R.string.wrong_email_or_password);
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: should write 
dlgAlert.setMessage
.setTitle
.setPositiveButton
.setCancelable
setPositiveButton
.create
.show

look better

File: app/src/main/java/com/framgia/forder/screen/login/LoginActivity.java [hunk 12]
```diff
@@ -17,16 +25,20 @@
...
-        LoginContract.Presenter presenter = new LoginPresenter(mViewModel);
+        mViewModel = new LoginViewModel(this, user, navigator);
+        UserRepository userRepository =
+                new UserRepository(new UserRemoteDataSource(FOrderServiceClient.getInstance()),
+                        new UserLocalDataSource(realmApi, prefsApi));
+        LoginContract.Presenter presenter = new LoginPresenter(mViewModel, userRepository);
         mViewModel.setPresenter(presenter);
 
-        ActivityLoginBinding binding =
+        ActivityLoginBinding mbinding =
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/forder/screen/login/LoginActivity.java]: check convention

File: app/src/main/java/com/framgia/forder/data/source/local/UserLocalDataSource.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    private RealmApi mRealmApi;
+    private SharedPrefsApi mSharedPrefsApi;
+
+    public UserLocalDataSource(@NonNull RealmApi realmApi, @NonNull SharedPrefsApi prefsApi) {
+        mRealmApi = realmApi;
+        mSharedPrefsApi = prefsApi;
+    }
+
+    public void saveUser(User user) {
+        mSharedPrefsApi.put(SharedPrefsApi.USER_KEY, user);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/UserLocalDataSource.java]: change to 
```java
String data =  new Gson().toJson(user)
mSharedPrefsApi.put(SharedPrefsApi.KEY_USER, data);
```

File: app/src/main/java/com/framgia/forder/data/source/local/sharedprf/SharedPrefsApi.java [hunk 14]
```diff
@@ -5,6 +5,7 @@
  */
 
 public interface SharedPrefsApi {
+    public static final String USER_KEY = "USER_KEY";
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/sharedprf/SharedPrefsApi.java]: move to ``SharedPrefsKey``

File: app/src/main/java/com/framgia/forder/screen/login/LoginActivity.java (hunk 12 above)
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginActivity.java]: ? why change this name

File: app/src/main/java/com/framgia/forder/utils/navigator/Navigator.java [hunk 15]
```diff
@@ -45,15 +45,6 @@ public void startActivity(@NonNull Class<? extends Activity> clazz) {
         mActivity.startActivity(new Intent(mActivity, clazz));
     }
 
-    public void finishActivity() {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/utils/navigator/Navigator.java]: ? why u delete this code?

File: app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java [hunk 16]
```diff
@@ -1,16 +1,27 @@
...
+import rx.functions.Action1;
+import rx.schedulers.Schedulers;
+import rx.subscriptions.CompositeSubscription;
+
 /**
  * Listens to user actions from the UI ({@link LoginActivity}), retrieves the data and updates
  * the UI as required.
  */
 final class LoginPresenter implements LoginContract.Presenter {
-    private static final String TAG = LoginPresenter.class.getName();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java]: why delete this TAG, It use for debug

File: app/src/main/java/com/framgia/forder/data/source/UserRepository.java [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Created by levutantuan on 4/13/17.
+ */
+
+public class UserRepository {
+    private UserRemoteDataSource mUserRemoteDataSource;
+    private UserLocalDataSource mUserLocalDataSource;
+
+    public UserRepository(UserRemoteDataSource userRemoteDataSource,
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/UserRepository.java]: change to 
```java
public UserRepository(UserDataSource.RemoteDataSource userRemoteDataSource,
             UserDataSource.LocalDataSource userLocalDataSource) {
               .....................
                }
```

File: app/src/main/java/com/framgia/forder/data/source/local/UserLocalDataSource.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    private SharedPrefsApi mSharedPrefsApi;
+
+    public UserLocalDataSource(@NonNull RealmApi realmApi, @NonNull SharedPrefsApi prefsApi) {
+        mRealmApi = realmApi;
+        mSharedPrefsApi = prefsApi;
+    }
+
+    public void saveUser(User user) {
+        String data = new Gson().toJson(user);
+        mSharedPrefsApi.put(SharedPrefsKey.KEY_USER, user);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/UserLocalDataSource.java]: change to ``mSharedPrefsApi.put(SharedPrefsKey.KEY_USER, data);``

File: app/src/main/java/com/framgia/forder/data/source/local/UserLocalDataSource.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    public User getUser() {
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/UserLocalDataSource.java]: remove empty line

File: app/src/main/java/com/framgia/forder/screen/login/LoginActivity.java [hunk 20]
```diff
@@ -17,13 +25,16 @@
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
-
+        RealmApi realmApi = new RealmApi();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginActivity.java]: remove Realm API,
We will not use RealmDB in this PR

File: app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java [hunk 21]
```diff
@@ -19,5 +32,48 @@ public void onStart() {
...
+                .subscribe(new Action1<User>() {
+                    @Override
+                    public void call(User user) {
+                        mViewModel.onLoginSuccess();
+                    }
+                }, new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable throwable) {
+                        //Todo show dialog error later!
+                        mViewModel.onLoginError();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginPresenter.java]: change to ``mViewModel.onLoginError(throwable);`` and remove ``//Todo show dialog error later!``

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 22]
```diff
@@ -1,20 +1,53 @@
...
-    public LoginViewModel(Navigator navigator) {
+    public LoginViewModel(Context context, User user, Navigator navigator) {
+        mContext = context;
         mNavigator = navigator;
+        mUser = user;
+    }
+
+    @Override
+    public void onLoginClick() {
+
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: remove empty line

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 23]
```diff
@@ -32,9 +65,66 @@ public void setPresenter(LoginContract.Presenter presenter) {
...
+        Toast.makeText(mContext, R.string.login_error, Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+
+    public void onLoginSuccess() {
+        //TODO Login Success !
+        Toast.makeText(mContext, R.string.login_success, Toast.LENGTH_SHORT).show();
+        mPresenter.saveUser(mUser);
+        mNavigator.startActivity(ChooseDomainActivity.class);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: add ``mNavigator.finishActivity()`` in here

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 24]
```diff
@@ -1,20 +1,53 @@
...
     private LoginContract.Presenter mPresenter;
+    private String mUsernameError;
+    private String mPassswordError;
+    private String mUsername;
+    private String mPassword;
+    private User mUser = new User();
 
-    public LoginViewModel(Navigator navigator) {
+    public LoginViewModel(Context context, User user, Navigator navigator) {
+        mContext = context;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: remove context from Contructor of ViewModel

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 25]
```diff
@@ -32,9 +65,66 @@ public void setPresenter(LoginContract.Presenter presenter) {
...
     }
 
-    public void onLoginButtonClicked(View view) {
-        //TODO change later
-        mNavigator.startActivity(MainActivity.class);
-        mNavigator.finishActivity();
+    @Override
+    public void onLoginError(Throwable throwable) {
+        //TODO Login Error !
+        Toast.makeText(mContext, R.string.login_error, Toast.LENGTH_SHORT).show();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: Dialog

File: app/src/main/java/com/framgia/forder/data/source/remote/api/service/FOrderApi.java [hunk 26]
```diff
@@ -14,4 +16,7 @@
     @GET("/search/users")
     Observable<Void> searchGithubUsers(@Query("per_page") int limit,
             @Query("q") String searchTerm);
+    @GET("v1/authen_user_tokens")
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/remote/api/service/FOrderApi.java]: add enter in here.

File: app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java [hunk 27]
```diff
@@ -1,20 +1,48 @@
...
-
+public class LoginViewModel extends BaseObservable implements LoginContract.ViewModel {
+    public final ObservableField<Integer> progressBarVisibility = new ObservableField<>();
     private Navigator mNavigator;
     private LoginContract.Presenter mPresenter;
+    private String mUsernameError;
+    private String mPassswordError;
+    private String mUsername;
+    private String mPassword;
+    private User mUser = new User();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/login/LoginViewModel.java]: not init in here
change to
``private User mUser``

**Pull request merged**